[ iOS ] fast/mediastream/video-srcObject-fit-fill.html is a constant text failure

youennf opened the pull request:
#### 3943142d2ef1777fc8977e6158eb10e053c3f8ca
<pre>
[ iOS ] fast/mediastream/video-srcObject-fit-fill.html is a constant text failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=264293">https://bugs.webkit.org/show_bug.cgi?id=264293</a>
<a href="https://rdar.apple.com/118020922">rdar://118020922</a>

Reviewed by Jer Noble.

The WebAVPlayerLayer was not receiving the gravity from WebProcess on iOS, which breaks its bounds computation.
Update RemoteLayerTreePropertyApplier::applyPropertiesToLayer to take into account WKVideoView.

We update fast/mediastream/video-srcObject-fit-fill.html to check for aspect ratio instead of pixel width to handle the various simulators.

* LayoutTests/fast/mediastream/video-srcObject-fit-fill.html:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::createViewWithID):
* Source/WebKit/UIProcess/ios/WKVideoView.h:
* Source/WebKit/UIProcess/ios/WKVideoView.mm:
(-[WKVideoView initWithFrame:playerView:]):
(+[WKVideoView playerLayerFromView:]):

Canonical link: <a href="https://commits.webkit.org/271859@main">https://commits.webkit.org/271859@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42b74951e03c6dc697f22d5ca97cdfeae7c8202d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29213 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7884 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30553 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31802 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26570 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10025 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5189 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26574 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6545 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25029 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5642 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5785 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26066 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33140 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26682 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32010 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3933 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29798 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7431 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/25860 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7079 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6263 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6275 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -299,8 +299,14 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
+#if PLATFORM(IOS_FAMILY)
+        if (layerTreeNode)
+            playerLayer = layerTreeNode->uiView().subviews.firstObject.layer;
+#endif
```
This is very fragile. I suggest you add a check that the `[layerTreeNode->uiView() isKindOfClass:WKVideoView.class]`, add a `@property (retain, nonatomic) WebAVPlayerLayerView* playerLayerView` to WKVideoView (which adds the playerLayerView to it's view hierarchy), and query that property rather than relying on the playerLayer being the first subview's layer.

jernoble commented on Source/WebKit/UIProcess/ios/WKVideoView.mm:
```diff
@@ -33,6 +33,23 @@ + (Class)layerClass
+    ASSERT([view isKindOfClass:WKVideoView.class]);
+    return view.subviews.firstObject.layer;
+}
+
```
Making this a class method is super weird. `-isKindOfClass:` is not expensive; in the 99% case it'll be a pointer-equality check. We shouldn't be micro-optimizing to avoid it.

Also, this is still fragile, though less so. Just add a `@property (retain, nonatomic) WebAVPlayerLayerView *playerLayerView`, assign it in the -init method, and pull the layer directly out of it.

youennf committed:
[ iOS ] fast/mediastream/video-srcObject-fit-fill.html is a constant text failure
https://bugs.webkit.org/show_bug.cgi?id=264293
rdar://118020922

Reviewed by Jer Noble.

The WebAVPlayerLayer was not receiving the gravity from WebProcess on iOS, which breaks its bounds computation.
Update RemoteLayerTreePropertyApplier::applyPropertiesToLayer to take into account WKVideoView.

We update fast/mediastream/video-srcObject-fit-fill.html to check for aspect ratio instead of pixel width to handle the various simulators.

* LayoutTests/fast/mediastream/video-srcObject-fit-fill.html:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(WebKit::VideoPresentationManagerProxy::createViewWithID):
* Source/WebKit/UIProcess/ios/WKVideoView.h:
* Source/WebKit/UIProcess/ios/WKVideoView.mm:
(-[WKVideoView initWithFrame:playerView:]):
(+[WKVideoView playerLayerFromView:]):

Canonical link: https://commits.webkit.org/271859@main
